clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -349,7 +351,16 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
 
         if (numHeaders == 0)
             return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, Record.EMPTY_HEADERS);
+        try {
```

Comment:
The buffer underflow could happen above as well. Perhaps we can expand the `try` to cover this full method and make the exception more generic?